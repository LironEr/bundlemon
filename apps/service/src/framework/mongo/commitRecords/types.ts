import type { CommitRecordReview, Compression, CommitRecord } from 'bundlemon-utils';

export interface CommitRecordReviewDB extends Omit<CommitRecordReview, 'createdAt'> {
  createdAt: Date;
}

export interface AssetLimits {
  maxSize?: number;
  maxPercentIncrease?: number;
}

export interface AssetMatchCriteria {
  pattern: string;
  friendlyName?: string;
  compression: Compression;
  limits?: AssetLimits;
}

export interface AssetMatch {
  path: string;
  size: number;
}

export interface WatchedFileHits extends AssetMatchCriteria {
  matches: AssetMatch[];
}

export interface WatchedGroupHits extends AssetMatchCriteria, Omit<AssetMatch, 'path'> {}

export interface CommitRecordDB {
  projectId: string;
  creationDate: Date;
  subProject?: string;
  branch: string;
  commitSha: string;
  baseBranch?: string;
  prNumber?: string;
  commitMsg?: string;
  files?: WatchedFileHits[];
  groups?: WatchedGroupHits[];
  reviews?: CommitRecordReviewDB[];
  outputs?: CommitRecord['outputs'];
}
