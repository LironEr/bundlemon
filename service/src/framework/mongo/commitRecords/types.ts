import type { CommitRecordApprover, Compression, CommitRecord } from 'bundlemon-utils';

export interface CommitRecordApproverDB extends Omit<CommitRecordApprover, 'approveDate'> {
  approveDate: Date;
}

export interface AssetLimits {
  maxSize?: number;
  maxPercentIncrease?: number;
}

export interface AssetMatchCriteria {
  pattern: string;
  friendlyName?: string;
  compression: Compression;
  limits?: AssetLimits;
}

export interface AssetMatch {
  path: string;
  size: number;
}

export interface WatchedFileHits extends AssetMatchCriteria {
  matches: AssetMatch[];
}

export interface WatchedGroupHits extends AssetMatchCriteria, Omit<AssetMatch, 'path'> {}

export interface CommitRecordDB {
  projectId: string;
  creationDate: Date;
  subProject?: string;
  branch: string;
  commitSha: string;
  baseBranch?: string;
  prNumber?: string;
  commitMsg?: string;
  files?: WatchedFileHits[];
  groups?: WatchedGroupHits[];
  approvers?: CommitRecordApproverDB[];
  outputs?: CommitRecord['outputs'];
}
