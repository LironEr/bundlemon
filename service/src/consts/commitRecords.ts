export enum CommitRecordsQueryResolution {
  All = 'all',
  Days = 'days',
  Weeks = 'weeks',
  Months = 'months',
}

export enum BaseRecordCompareTo {
  PreviousCommit = 'PREVIOUS_COMMIT',
  LatestCommit = 'LATEST_COMMIT',
}

export enum CreateCommitRecordAuthType {
  GithubActions = 'GITHUB_ACTIONS',
}
