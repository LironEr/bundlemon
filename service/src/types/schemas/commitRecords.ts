/* istanbul ignore file */

import type { CommitRecordPayload } from 'bundlemon-utils';
import type { CommitRecordsQueryResolution } from '../../consts/commitRecords';
import type { BaseRequestSchema, BaseGetRequestSchema, AuthHeaders } from './common';

interface ProjectIdParams {
  /**
   * @pattern ^[0-9a-fA-F]{24}$
   */
  projectId: string;
}

export interface CreateCommitRecordRequestSchema extends BaseRequestSchema {
  body: CommitRecordPayload;
  params: ProjectIdParams;
  headers: AuthHeaders;
}

interface GetCommitRecordRequestParams extends ProjectIdParams {
  commitRecordId: string;
}

export interface GetCommitRecordRequestSchema extends BaseGetRequestSchema {
  params: GetCommitRecordRequestParams;
}

export interface GetCommitRecordsQuery {
  branch: string;
  latest?: boolean;
  resolution?: CommitRecordsQueryResolution;
  subProject?: string;
}

export interface GetCommitRecordsRequestSchema extends BaseGetRequestSchema {
  params: ProjectIdParams;
  query: GetCommitRecordsQuery;
}
