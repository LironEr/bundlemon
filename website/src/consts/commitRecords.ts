export enum CommitRecordsQueryResolution {
  All = 'all',
  Days = 'days',
  Weeks = 'weeks',
  Months = 'months',
}
