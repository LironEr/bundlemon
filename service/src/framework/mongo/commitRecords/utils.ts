import { WithId } from 'mongodb';
import { CommitRecord, CommitRecordPayload, FileDetails } from 'bundlemon-utils';
import { CommitRecordDB, WatchedFileHits, WatchedGroupHits } from './types';
import { MAX_COMMIT_MSG_LENGTH } from '@/consts/commitRecords';
import { truncateString } from '@/utils/reportUtils';

export const filesToWatchedFileHits = (arr: FileDetails[]): WatchedFileHits[] => {
  const obj: Record<string, WatchedFileHits> = {};

  arr.forEach((f) => {
    const { path, size, maxSize, maxPercentIncrease, ...restDetails } = f;

    if (!obj[f.pattern]) {
      obj[f.pattern] = { ...restDetails, matches: [] };

      if (maxSize || maxPercentIncrease) {
        obj[f.pattern].limits = {
          maxSize,
          maxPercentIncrease,
        };
        removeUndefinedFromObject(obj[f.pattern].limits);
      }
    }

    obj[f.pattern].matches.push({ path, size });
  });

  return Object.values(obj);
};

export const groupsToWatchedGroupHits = (arr: FileDetails[]): WatchedGroupHits[] => {
  const obj: Record<string, WatchedGroupHits> = {};

  arr.forEach((f) => {
    const { maxSize, maxPercentIncrease, path, ...restDetails } = f;

    obj[f.pattern] = { ...restDetails };

    if (maxSize || maxPercentIncrease) {
      obj[f.pattern].limits = {
        maxSize,
        maxPercentIncrease,
      };
      removeUndefinedFromObject(obj[f.pattern].limits);
    }
  });

  return Object.values(obj);
};

export const watchedFileHitsToFiles = (hits: WatchedFileHits[] | undefined): FileDetails[] => {
  if (!hits) {
    return [];
  }

  const files: FileDetails[] = [];

  hits.forEach((file) => {
    const { limits, matches, ...rest } = file;

    const base: Omit<FileDetails, 'path' | 'size'> = {
      ...rest,
      ...limits,
    };

    matches.forEach((match) => {
      files.push({ ...base, ...match });
    });
  });

  return files;
};

export const watchedGroupHitsToGroups = (hits: WatchedGroupHits[] | undefined): FileDetails[] => {
  if (!hits) {
    return [];
  }

  const files: FileDetails[] = [];

  hits.forEach((file) => {
    const { limits, ...rest } = file;

    files.push({ ...limits, ...rest, path: rest.pattern });
  });

  return files;
};

export const commitRecordDBToResponse = (record: WithId<CommitRecordDB>): CommitRecord => {
  const { _id, creationDate, files, groups, reviews, ...restRecord } = record;

  const response: CommitRecord = {
    id: _id.toHexString(),
    creationDate: creationDate.toISOString(),
    files: watchedFileHitsToFiles(files),
    groups: watchedGroupHitsToGroups(groups),
    ...restRecord,
  };

  if (reviews) {
    response.reviews = reviews.map((r) => ({
      ...r,
      createdAt: r.createdAt.toISOString(),
    }));
  }

  return response;
};

export const commitRecordPayloadToDBModel = (
  projectId: string,
  payload: CommitRecordPayload
): Omit<CommitRecordDB, '_id'> => {
  const { files, groups, commitMsg, ...restPayload } = payload;
  const recordToSave: Omit<CommitRecordDB, '_id'> = {
    ...restPayload,
    projectId,
    creationDate: new Date(),
  };

  if (commitMsg) {
    recordToSave.commitMsg = truncateString(commitMsg, MAX_COMMIT_MSG_LENGTH);
  }

  if (payload.files && payload.files.length > 0) {
    recordToSave.files = filesToWatchedFileHits(payload.files);
  }

  if (payload.groups && payload.groups.length > 0) {
    recordToSave.groups = groupsToWatchedGroupHits(payload.groups);
  }

  return recordToSave;
};

function removeUndefinedFromObject(obj: any) {
  Object.keys(obj).forEach((key) => {
    if (obj[key] === undefined) {
      delete obj[key];
    }
  });
}
