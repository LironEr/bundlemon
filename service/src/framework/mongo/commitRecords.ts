import { ObjectId, WithId, ReturnDocument, Filter } from 'mongodb';
import { BaseRecordCompareTo, CommitRecordsQueryResolution, MAX_COMMIT_MSG_LENGTH } from '../../consts/commitRecords';
import { getCollection } from './client';
import { truncateString } from '../../utils/reportUtils';

import type { CommitRecordPayload, CommitRecord } from 'bundlemon-utils';
import type { GetCommitRecordsQuery } from '../../types/schemas';

interface CommitRecordDB extends CommitRecordPayload {
  projectId: string;
  creationDate: Date;
}

export const getCommitRecordsCollection = () => getCollection<CommitRecordDB>('commitRecords');

const commitRecordDBToResponse = (record: WithId<CommitRecordDB>): CommitRecord => {
  const { _id, creationDate, ...restRecord } = record;

  return { id: _id.toHexString(), creationDate: creationDate.toISOString(), ...restRecord };
};

export const createCommitRecord = async (projectId: string, record: CommitRecordPayload): Promise<CommitRecord> => {
  const commitRecordsCollection = await getCommitRecordsCollection();
  const recordToSave: Omit<CommitRecordDB, '_id'> = {
    ...record,
    projectId,
    creationDate: new Date(),
  };

  if (recordToSave.commitMsg) {
    recordToSave.commitMsg = truncateString(recordToSave.commitMsg, MAX_COMMIT_MSG_LENGTH);
  }

  const result = await commitRecordsCollection.findOneAndReplace(
    { projectId, subProject: record.subProject, commitSha: record.commitSha },
    recordToSave,
    {
      upsert: true,
      returnDocument: ReturnDocument.AFTER,
    }
  );

  const newRecord = result.value;

  if (!newRecord) {
    throw new Error('Failed to findOneAndReplace record');
  }

  return commitRecordDBToResponse(newRecord);
};

interface GetCommitRecordParams {
  projectId: string;
  commitRecordId: string;
}

export const getCommitRecord = async ({
  projectId,
  commitRecordId,
}: GetCommitRecordParams): Promise<CommitRecord | undefined> => {
  const commitRecordsCollection = await getCommitRecordsCollection();
  const record = await commitRecordsCollection.findOne<WithId<CommitRecordDB>>({
    _id: new ObjectId(commitRecordId),
    projectId,
  });

  if (!record) {
    return undefined;
  }

  return commitRecordDBToResponse(record);
};

const MAX_RECORDS = 100;

const resolutions: Record<
  Exclude<CommitRecordsQueryResolution, CommitRecordsQueryResolution.All>,
  Record<string, any>
> = {
  [CommitRecordsQueryResolution.Days]: {
    year: {
      $year: '$creationDate',
    },
    month: {
      $month: '$creationDate',
    },
    day: {
      $dayOfMonth: '$creationDate',
    },
  },
  [CommitRecordsQueryResolution.Weeks]: {
    year: {
      $year: '$creationDate',
    },
    week: {
      $week: '$creationDate',
    },
  },
  [CommitRecordsQueryResolution.Months]: {
    year: {
      $year: '$creationDate',
    },
    month: {
      $month: '$creationDate',
    },
  },
};

export async function getCommitRecords(
  projectId: string,
  { branch, latest, resolution, subProject, olderThan }: GetCommitRecordsQuery
): Promise<CommitRecord[]> {
  const commitRecordsCollection = await getCommitRecordsCollection();

  let creationDateFilter: Filter<Pick<CommitRecordDB, 'creationDate'>> | undefined = undefined;

  if (olderThan) {
    creationDateFilter = { creationDate: { $lt: olderThan } };
  }

  let records: WithId<CommitRecordDB>[] = [];

  if (resolution && resolution !== CommitRecordsQueryResolution.All) {
    records = await commitRecordsCollection
      .aggregate<WithId<CommitRecordDB>>([
        {
          $match: {
            ...creationDateFilter,
            projectId,
            branch,
            subProject,
          },
        },
        {
          $sort: {
            creationDate: -1,
          },
        },
        {
          $group: {
            _id: resolutions[resolution],
            doc: {
              $first: '$$ROOT',
            },
          },
        },
        {
          $replaceRoot: {
            newRoot: '$doc',
          },
        },
        {
          $sort: {
            creationDate: -1,
          },
        },
        {
          $limit: latest ? 1 : MAX_RECORDS,
        },
      ])
      .toArray();
  } else {
    records = await commitRecordsCollection
      .find(
        { ...creationDateFilter, projectId, branch: branch, subProject },
        { sort: { creationDate: -1 }, limit: latest ? 1 : MAX_RECORDS }
      )
      .toArray();
  }

  return records.map(commitRecordDBToResponse);
}

export async function getSubprojects(projectId: string) {
  const commitRecordsCollection = await getCommitRecordsCollection();
  const subProjects = await commitRecordsCollection.distinct('subProject', { projectId });

  return subProjects.filter((s) => !!s);
}

export interface CommitRecordWithBase {
  record: CommitRecord;
  baseRecord?: CommitRecord;
}

export async function getCommitRecordWithBase(
  { projectId, commitRecordId }: GetCommitRecordParams,
  compareTo = BaseRecordCompareTo.PreviousCommit
): Promise<CommitRecordWithBase | undefined> {
  const record = await getCommitRecord({ projectId, commitRecordId });

  if (!record) {
    return undefined;
  }

  const baseRecord = (
    await getCommitRecords(projectId, {
      branch: record.baseBranch ?? record.branch,
      subProject: record.subProject,
      latest: true,
      olderThan: compareTo === BaseRecordCompareTo.PreviousCommit ? new Date(record.creationDate) : undefined,
    })
  )?.[0];

  return { record, baseRecord };
}
