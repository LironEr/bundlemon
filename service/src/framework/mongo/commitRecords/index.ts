import { ObjectId, WithId, ReturnDocument, Filter } from 'mongodb';
import { BaseRecordCompareTo, CommitRecordsQueryResolution, MAX_QUERY_RECORDS } from '@/consts/commitRecords';
import { getCollection } from '@/framework/mongo/client';
import { commitRecordDBToResponse, commitRecordPayloadToDBModel } from './utils';

import type { CommitRecordPayload, CommitRecord, CommitRecordReview, CommitRecordGitHubOutputs } from 'bundlemon-utils';
import type { GetCommitRecordsQuery } from '@/types/schemas';
import type { CommitRecordDB, CommitRecordReviewDB } from './types';

export const getCommitRecordsCollection = () => getCollection<CommitRecordDB>('commitRecords');

export const createCommitRecord = async (projectId: string, payload: CommitRecordPayload): Promise<CommitRecord> => {
  const commitRecordsCollection = await getCommitRecordsCollection();
  const recordToSave = commitRecordPayloadToDBModel(projectId, payload);

  const newRecord = await commitRecordsCollection.findOneAndReplace(
    { projectId, subProject: payload.subProject, commitSha: payload.commitSha },
    recordToSave,
    {
      upsert: true,
      returnDocument: ReturnDocument.AFTER,
    }
  );

  if (!newRecord) {
    throw new Error('Failed to findOneAndReplace record');
  }

  return commitRecordDBToResponse(newRecord);
};

interface GetCommitRecordParams {
  projectId: string;
  commitRecordId: string;
}

export const getCommitRecord = async ({
  projectId,
  commitRecordId,
}: GetCommitRecordParams): Promise<CommitRecord | undefined> => {
  const commitRecordsCollection = await getCommitRecordsCollection();
  const record = await commitRecordsCollection.findOne<WithId<CommitRecordDB>>({
    _id: new ObjectId(commitRecordId),
    projectId,
  });

  if (!record) {
    return undefined;
  }

  return commitRecordDBToResponse(record);
};

const resolutions: Record<
  Exclude<CommitRecordsQueryResolution, CommitRecordsQueryResolution.All>,
  Record<string, any>
> = {
  [CommitRecordsQueryResolution.Days]: {
    year: {
      $year: '$creationDate',
    },
    month: {
      $month: '$creationDate',
    },
    day: {
      $dayOfMonth: '$creationDate',
    },
  },
  [CommitRecordsQueryResolution.Weeks]: {
    year: {
      $year: '$creationDate',
    },
    week: {
      $week: '$creationDate',
    },
  },
  [CommitRecordsQueryResolution.Months]: {
    year: {
      $year: '$creationDate',
    },
    month: {
      $month: '$creationDate',
    },
  },
};

export async function getCommitRecords(
  projectId: string,
  { branch, latest, resolution, subProject, olderThan }: GetCommitRecordsQuery
): Promise<CommitRecord[]> {
  const commitRecordsCollection = await getCommitRecordsCollection();

  let creationDateFilter: Filter<Pick<CommitRecordDB, 'creationDate'>> | undefined = undefined;

  if (olderThan) {
    creationDateFilter = { creationDate: { $lt: olderThan } };
  }

  let records: WithId<CommitRecordDB>[] = [];

  if (resolution && resolution !== CommitRecordsQueryResolution.All) {
    records = await commitRecordsCollection
      .aggregate<WithId<CommitRecordDB>>([
        {
          $match: {
            ...creationDateFilter,
            projectId,
            branch,
            subProject,
          },
        },
        {
          $sort: {
            creationDate: -1,
          },
        },
        {
          $group: {
            _id: resolutions[resolution],
            doc: {
              $first: '$$ROOT',
            },
          },
        },
        {
          $replaceRoot: {
            newRoot: '$doc',
          },
        },
        {
          $sort: {
            creationDate: -1,
          },
        },
        {
          $limit: latest ? 1 : MAX_QUERY_RECORDS,
        },
      ])
      .toArray();
  } else {
    records = await commitRecordsCollection
      .find(
        { ...creationDateFilter, projectId, branch: branch, subProject },
        { sort: { creationDate: -1 }, limit: latest ? 1 : MAX_QUERY_RECORDS }
      )
      .toArray();
  }

  return records.map(commitRecordDBToResponse);
}

export async function getSubprojects(projectId: string) {
  const commitRecordsCollection = await getCommitRecordsCollection();
  const subProjects = await commitRecordsCollection.distinct('subProject', { projectId });

  return subProjects.filter((s) => !!s);
}

export interface CommitRecordWithBase {
  record: CommitRecord;
  baseRecord?: CommitRecord;
}

export async function getCommitRecordWithBase(
  query: GetCommitRecordParams | { projectId: string; record: CommitRecord },
  compareTo = BaseRecordCompareTo.PreviousCommit
): Promise<CommitRecordWithBase | undefined> {
  const { projectId } = query;

  let record: CommitRecord | undefined = undefined;

  if ('record' in query) {
    record = query.record;
  } else {
    record = await getCommitRecord({ projectId, commitRecordId: query.commitRecordId });

    if (!record) {
      return undefined;
    }
  }

  const baseRecord = (
    await getCommitRecords(projectId, {
      branch: record.baseBranch ?? record.branch,
      subProject: record.subProject,
      latest: true,
      olderThan: compareTo === BaseRecordCompareTo.PreviousCommit ? new Date(record.creationDate) : undefined,
    })
  )?.[0];

  return { record, baseRecord };
}

interface GetCommitRecordsWithBaseByCommitShaQuery {
  commitSha: string;
  prNumber: number;
}

export async function getCommitRecordsWithBaseByCommitSha(
  projectId: string,
  { prNumber, commitSha }: GetCommitRecordsWithBaseByCommitShaQuery
): Promise<CommitRecordWithBase[]> {
  const commitRecordsCollection = await getCommitRecordsCollection();
  const records = (
    await commitRecordsCollection.find({ projectId, prNumber: String(prNumber), commitSha }).toArray()
  ).map(commitRecordDBToResponse);

  const recordsWithBase = (await Promise.all(
    records.map(async (record) => getCommitRecordWithBase({ projectId, record }))
  )) as CommitRecordWithBase[];

  return recordsWithBase;
}

export async function addReviewToCommitRecord(
  projectId: string,
  commitRecordId: string,
  review: CommitRecordReview
): Promise<CommitRecord> {
  const commitRecordsCollection = await getCommitRecordsCollection();

  const result = await commitRecordsCollection.findOneAndUpdate(
    { projectId, _id: new ObjectId(commitRecordId) },
    { $push: { reviews: transformReviewToDB(review) } },
    { returnDocument: ReturnDocument.AFTER }
  );

  if (!result) {
    throw new Error('Failed to update reviews list');
  }

  return commitRecordDBToResponse(result);
}

function transformReviewToDB(review: CommitRecordReview | undefined): CommitRecordReviewDB | undefined {
  if (!review) {
    return undefined;
  }

  const { createdAt, ...rest } = review;

  return {
    ...rest,
    createdAt: new Date(createdAt),
  };
}

export async function setCommitRecordGithubOutputs(
  projectId: string,
  commitRecordId: string,
  outputs: CommitRecordGitHubOutputs
): Promise<void> {
  const commitRecordsCollection = await getCommitRecordsCollection();

  await commitRecordsCollection.findOneAndUpdate(
    { projectId, _id: new ObjectId(commitRecordId) },
    { $set: { 'outputs.github': outputs } }
  );
}
