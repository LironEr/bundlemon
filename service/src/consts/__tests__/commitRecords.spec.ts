import * as consts from '../commitRecords';

describe('consts', () => {
  test('snapshot', () => {
    expect(consts).toMatchSnapshot();
  });
});
