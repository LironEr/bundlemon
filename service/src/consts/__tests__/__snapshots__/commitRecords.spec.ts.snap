// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`consts snapshot 1`] = `
Object {
  "BaseRecordCompareTo": Object {
    "LatestCommit": "LATEST_COMMIT",
    "PreviousCommit": "PREVIOUS_COMMIT",
  },
  "CommitRecordsQueryResolution": Object {
    "All": "all",
    "Days": "days",
    "Months": "months",
    "Weeks": "weeks",
  },
  "CreateCommitRecordAuthType": Object {
    "GithubActions": "GITHUB_ACTIONS",
  },
}
`;
